### PR TITLE
try to add a better localhost resolver. See issue #44

### PR DESCRIPTION
**Pull request by lassic**
A "Smarter" localhost resolver:

see issue #44 
Trying to avoid a lot of calls to java.net.InetAddress.getLocalHost()
which under AWS seem to trigger DNS resolving and have non-zero (~10ms) latency _per event_.
Usually, the hostname doesn't change so often to warrant a real query.

A real call to java.net.InetAddress.getLocalHost().getHostName()
is made only if:
1) the refresh interval has passed (=result is stale)
AND
2) no env vars that identify the hostname are found

### CODE REVIEW

File: riemann-java-client.iml [hunk 1]
```diff
@@ -6,23 +6,25 @@
     <content url="file://$MODULE_DIR$">
       <sourceFolder url="file://$MODULE_DIR$/src/main/java" isTestSource="false" />
```
Comment (itaifrenkel) [inline riemann-java-client.iml:7]: are these changes related to this feature?

File: src/main/java/com/aphyr/riemann/client/LocalhostResolver.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+ * 2) no env vars that identify the hostname are found
+ */
+public class LocalhostResolver {
+
+    // default hostname env var names on Win/Nix
+    public static final String COMPUTERNAME = "COMPUTERNAME"; // Windows
+    public static final String HOSTNAME = "HOSTNAME"; // Nix
+
+    // how often should we refresh the cached hostname
+    public static long RefreshIntervalMillis = 60 * 1000;
```
Comment (itaifrenkel) [inline src/main/java/com/aphyr/riemann/client/LocalhostResolver.java]: casing (first letter lowecase)

File: src/main/java/com/aphyr/riemann/client/LocalhostResolver.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    // enables setting a custom env var used for resolving
+    public static String CustomEnvVarName = null;
```
Comment (itaifrenkel) [inline src/main/java/com/aphyr/riemann/client/LocalhostResolver.java]: Too much customization. Would you ever need this?

File: src/main/java/com/aphyr/riemann/client/LocalhostResolver.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+     */
+    public static void refreshResolve() {
+        try {
+            hostname = resolveByEnv();
+            if(hostname == null || hostname.isEmpty()) {
+                hostname = java.net.InetAddress.getLocalHost().getHostName();
+                lastNetUpdate = System.currentTimeMillis();
+            }
+        } catch (UnknownHostException e) {
+            //e.printStackTrace();
```
Comment (itaifrenkel) [inline src/main/java/com/aphyr/riemann/client/LocalhostResolver.java]: replace comment with // fallthrough

File: src/main/java/com/aphyr/riemann/client/LocalhostResolver.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+        return hostname;
+    }
+
+    /**
+     * forces a new resolve even if refresh interval has not passed yet
+     */
+    public static void refreshResolve() {
+        try {
+            hostname = resolveByEnv();
```
Comment (itaifrenkel) [inline src/main/java/com/aphyr/riemann/client/LocalhostResolver.java]: this shouldn't be here. There is no need to resolve the environment variable again and again. There is also no need to call System.currentTimeMillis() each time in this case.

File: src/main/java/com/aphyr/riemann/client/LocalhostResolver.java [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+            if(hostname == null || hostname.isEmpty()) {
+                hostname = java.net.InetAddress.getLocalHost().getHostName();
+                lastNetUpdate = System.currentTimeMillis();
+            }
+        } catch (UnknownHostException e) {
+            //e.printStackTrace();
+        }
+        finally {
+            lastUpdate = System.currentTimeMillis();
```
Comment (itaifrenkel) [inline src/main/java/com/aphyr/riemann/client/LocalhostResolver.java]: you don't need this variable.

File: src/main/java/com/aphyr/riemann/client/LocalhostResolver.java [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+        }
+    }
+
+    /**
+     * try to resolve the hostname by env vars
+     *
+     * @return
+     */
+    private static String resolveByEnv() {
+        if(CustomEnvVarName != null) {
```
Comment (itaifrenkel) [inline src/main/java/com/aphyr/riemann/client/LocalhostResolver.java]: this is not really needed.

File: src/test/java/riemann/java/client/tests/LocalhostResolveTest.java [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+public class LocalhostResolveTest {
+
+    protected Map<String, String> env = new HashMap<String, String>();
+
+    protected static final String OS_NAME_PROP = "os.name";
+    protected static final String ExpectedWinHostname = "LocalHostResolverTestWin";
```
Comment (itaifrenkel) [inline src/test/java/riemann/java/client/tests/LocalhostResolveTest.java]: first letter must be lowercase

Comment (itaifrenkel) [inline src/test/java/riemann/java/client/tests/LocalhostResolveTest.java]: change everything to private. no need for protected

File: src/test/java/riemann/java/client/tests/LocalhostResolveTest.java [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+    }
+
+    @Test
+    public void testUpdateInterval() {
+        Assert.assertEquals(0, LocalhostResolver.getLastUpdateTime());
+        String hostname = LocalhostResolver.getResolvedHostname();
+        long lastUpdateTime = LocalhostResolver.getLastUpdateTime();
+        Assert.assertNotNull(hostname);
+        try {
+            Thread.sleep(1500);
```
Comment (itaifrenkel) [inline src/test/java/riemann/java/client/tests/LocalhostResolveTest.java]: you shouldn't use sleep in unit tests. You can expose get/setLastNetUpdateTime() for testing.
You can have a unit test to check that it actually changed the first time called, but did not change the second time called. Then inject a bogus lastNetUpdateTime call again and check that it was updated again.

File: src/test/java/riemann/java/client/tests/LocalhostResolveTest.java [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+                        Field field = cl.getDeclaredField("m");
+                        field.setAccessible(true);
+                        Object obj = field.get(env);
+                        Map<String, String> map = (Map<String, String>) obj;
+                        map.clear();
+                        map.putAll(newEnv);
+                    }
+                }
+            } catch (Exception e2) {
+                e2.printStackTrace();
```
Comment (itaifrenkel) [inline src/test/java/riemann/java/client/tests/LocalhostResolveTest.java]: throw. do not swallow exceptions

File: src/test/java/riemann/java/client/tests/LocalhostResolveTest.java [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+            }
+        } catch (Exception e1) {
+            e1.printStackTrace();
```
Comment (itaifrenkel) [inline src/test/java/riemann/java/client/tests/LocalhostResolveTest.java]: throw. do not swallow exceptions

File: src/test/java/riemann/java/client/tests/LocalhostResolveTest.java [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+
+
+    /**
+     * evil hack for testing (only!) with env var in-memory modification
+     * see: http://stackoverflow.com/a/7201825/1469004
+     *
+     * @param newEnv - to set in memory
+     */
+    protected static void setEnv(Map<String, String> newEnv) {
+        try {
```
Comment (itaifrenkel) [inline src/test/java/riemann/java/client/tests/LocalhostResolveTest.java:109]: why??? Just add setEnv() in the production code and document it's a testing hook (which overrides getEnv()).

File: src/main/java/com/aphyr/riemann/client/LocalhostResolver.java [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+            return System.getenv(CustomEnvVarName);
+        }
+
+        if(System.getProperty("os.name").startsWith("Windows")) {
+            return System.getenv(COMPUTERNAME);
```
Comment (itaifrenkel) [inline src/main/java/com/aphyr/riemann/client/LocalhostResolver.java]: use getenv instead of System.getenv which inludes an override for testing

File: riemann-java-client.iml (hunk 1 above)
Comment (lassic) [inline riemann-java-client.iml:7]: probably not, possibly new intellij version did this when importing the maven project.
otherwise it wouldn't recognize imports etc.

File: src/main/java/com/aphyr/riemann/client/LocalhostResolver.java (hunk 3 above)
Comment (lassic) [inline src/main/java/com/aphyr/riemann/client/LocalhostResolver.java]: I thought for testing/staging or in cases where you couldn't change the usually reserved env vars.

File: src/main/java/com/aphyr/riemann/client/LocalhostResolver.java (hunk 5 above)
Comment (lassic) [inline src/main/java/com/aphyr/riemann/client/LocalhostResolver.java]: hmm, but then how do you give a chance to get env var changes?
i also thought this mimics the closure code, unless I misread it (which is totally possible since I never coded in closure)

File: src/test/java/riemann/java/client/tests/LocalhostResolveTest.java (hunk 10 above)
Comment (lassic) [inline src/test/java/riemann/java/client/tests/LocalhostResolveTest.java]: but then I have to make every test method declare throw (which I prefer not to here, since it's a hacky test method anyway).
Can I use a RuntimeException?

File: src/test/java/riemann/java/client/tests/LocalhostResolveTest.java (hunk 12 above)
Comment (lassic) [inline src/test/java/riemann/java/client/tests/LocalhostResolveTest.java:109]: Let's talk about this, not sure why you prefer this approach.
I like to restrict hacks for testing to the test code and not actual code as much as possible.

File: src/main/java/com/aphyr/riemann/client/LocalhostResolver.java (hunk 13 above)
Comment (lassic) [inline src/main/java/com/aphyr/riemann/client/LocalhostResolver.java]: Not sure what you mean, I assume it relates to setEnv. Let's talk about the approach.

File: riemann-java-client.iml (hunk 1 above)
Comment (itaifrenkel) [inline riemann-java-client.iml:7]: ok. so please consider reverting these file changes

File: src/main/java/com/aphyr/riemann/client/LocalhostResolver.java (hunk 3 above)
Comment (itaifrenkel) [inline src/main/java/com/aphyr/riemann/client/LocalhostResolver.java]: You could do that with custom injection points in the dev code. Anyway, I think it's not a requirement now.

File: src/main/java/com/aphyr/riemann/client/LocalhostResolver.java (hunk 13 above)
Comment (itaifrenkel) [inline src/main/java/com/aphyr/riemann/client/LocalhostResolver.java]: As we talked, you suggested that adding an injection point in the dev code is not needed since we have the test code that can inject directly into env variables.

File: src/test/java/riemann/java/client/tests/LocalhostResolveTest.java (hunk 10 above)
Comment (itaifrenkel) [inline src/test/java/riemann/java/client/tests/LocalhostResolveTest.java]: If you have Guava in the pom use  throw Throwables.propagate(e2) otherwise use throw new RuntimeException(e2)

**Comment by lassic** [comment]
@itaifrenkel I hope I didn't forget anything, thanks.

**Comment by lassic** [comment]
Opening new PR for submission.
